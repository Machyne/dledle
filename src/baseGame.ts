// Give these string values so they can be round trip through being used as object keys.
// These are never encoded as-is so strings are fine.
export enum GameScore {
  Win = "Win",
  NearWin = "NearWin",
  Loss = "Loss",
}

export type GameResult = {
  score: GameScore;
  serializedResult: string;
};

export abstract class BaseGame {
  private _resultRegex: RegExp | null = null;

  /**
   * The name of the game.
   */
  abstract get name(): string;

  /**
   * The name of the game in lowercase with only alphanumeric characters.
   */
  get fileName(): string {
    return this.name.toLowerCase().replace(/[^a-z0-9]/g, "");
  }

  /**
   * A regex that matches the result of the game.
   */
  get resultRegex(): RegExp {
    if (this._resultRegex === null) {
      this._resultRegex = this._buildResultRegex();
    }
    return this._resultRegex;
  }

  /**
   * Builds a regex that matches the result of the game. The return value will be cached.
   */
  abstract _buildResultRegex(): RegExp;

  /**
   * A link to the game.
   */
  abstract get link(): string;

  /**
   * Parses the user-supplied game result and returns a serialized version of
   * the game data in addition to the "score".
   *
   * @param gameResult the result of the game run through the resultRegex
   */
  abstract serializeResult(gameResult: RegExpMatchArray): GameResult;

  /**
   * The share text of the result of the game given the serialized game data.
   *
   * @param serializedResult the serialized game data
   */
  abstract deserialize(serializedResult: string): string;
}
