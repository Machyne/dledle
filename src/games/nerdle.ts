import { BaseGame, GameScore } from "../baseGame";
import { DEFAULT_WORD_WIDTH, intToBase64, nextInt } from "../util/base64";
import { emojiSquares, emojiToRegexUnion, splitEmojiLines } from "../util/emoji";
import { b64ToEmojiRows, emojiRowsToB64 } from "../util/gameHelpers";

const rowsPerB64Word = 2;
const b64WordWidth = 5;

const squareValues = {
  [emojiSquares.black]: 0,
  [emojiSquares.white]: 0,
  [emojiSquares.purple]: 1,
  [emojiSquares.green]: 2,
};
const squareList = [emojiSquares.black, emojiSquares.purple, emojiSquares.green];

export class Nerdle extends BaseGame {
  get name(): string {
    return "Nerdle";
  }

  get link(): string {
    return "https://nerdlegame.com/";
  }

  _buildResultRegex(): RegExp {
    return new RegExp(
      `nerdlegame (\\d+) ([1-6X])/6\\s+((?:${emojiToRegexUnion(squareList)}{8}\\s*){1,6})`,
      "u",
    );
  }

  serializeResult(gameResult: RegExpMatchArray) {
    const [, gameNumber, gameScore, guessesBlock] = gameResult;
    const isWin = gameScore !== "X";
    const numGuesses = isWin ? parseInt(gameScore) : 6;
    // First encode the game number, win-or-not, and star-or-not.
    let encoded = intToBase64(2 * parseInt(gameNumber) + (isWin ? 1 : 0));
    // Next encode the guesses.
    const guesses = splitEmojiLines(guessesBlock);
    if (guesses.length !== numGuesses) {
      throw new Error(`Invalid number of guesses: ${guesses.length}`);
    }
    encoded += emojiRowsToB64({
      emojiLines: guesses,
      values: squareValues,
      rowsPerB64Word,
      b64WordWidth,
    });
    const score = isWin
      ? GameScore.Win
      : guesses.some((row) => row.filter((char) => char === emojiSquares.green).length == 4)
        ? GameScore.NearWin
        : GameScore.Loss;
    return { score, serializedResult: encoded };
  }

  deserialize(serializedResult: string) {
    if (
      serializedResult.length < DEFAULT_WORD_WIDTH + b64WordWidth ||
      (serializedResult.length - DEFAULT_WORD_WIDTH) % b64WordWidth !== 0
    ) {
      throw new Error(`Invalid Nerdle result: ${serializedResult}`);
    }
    let intResult: number;
    [intResult, serializedResult] = nextInt(serializedResult);
    const isWin = intResult & 1;
    const gameNumber = Math.floor(intResult / 2);

    const { rows: guesses, serializedResult: sr } = b64ToEmojiRows({
      serializedResult,
      lookup: squareList,
      rowLength: 8,
      rowsPerB64Word,
      b64WordWidth,
    });
    serializedResult = sr;
    return `nerdlegame ${gameNumber} ${isWin ? guesses.length.toString() : "X"}/6\n\n${guesses.join(
      "\n",
    )}`;
  }
}
