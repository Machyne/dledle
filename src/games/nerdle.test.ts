import { GameScore } from "../baseGame";
import { Nerdle } from "./nerdle";

const sampleEvenRowWin = `
nerdlegame 710 2/6

â¬›ðŸŸªðŸŸªâ¬›ðŸŸªâ¬›â¬›ðŸŸ©
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©
`;

const sampleOddRowWin = `
nerdlegame 711 3/6

â¬›ðŸŸªâ¬›â¬›ðŸŸªâ¬›â¬›ðŸŸ©
ðŸŸ©â¬›ðŸŸ©ðŸŸªâ¬›â¬›ðŸŸ©ðŸŸ©
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©
`;

// TODO add a sample loss and near-win? Seems like nerdle doesn't allow sharing without a win.

describe("Nerdle", () => {
  const nerdle = new Nerdle();

  for (const [name, input, expectedScore] of [
    ["sampleEvenRowWin", sampleEvenRowWin, GameScore.Win],
    ["sampleOddRowWin", sampleOddRowWin, GameScore.Win],
  ] as Array<[string, string, GameScore]>) {
    it(`should serialize and deserialize ${name} correctly`, () => {
      const match = nerdle.resultRegex.exec(input);
      expect(match).not.toBeNull();
      const { score, serializedResult } = nerdle.serializeResult(match!);

      expect(score).toEqual(expectedScore);
      expect(nerdle.deserialize(serializedResult)).toEqual(input.trim());
    });
  }
});
