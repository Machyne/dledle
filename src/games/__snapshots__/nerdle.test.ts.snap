// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nerdle serialized data should match snapshot 1`] = `
{
  "sampleEvenRowWin": {
    "score": "Win",
    "serializedResult": "NWAn4ZIF",
  },
  "sampleEvenRowWin trimmed": {
    "score": "Win",
    "serializedResult": "NWAn4ZIF",
  },
  "sampleEvenRowWin with extra link text": {
    "score": "Win",
    "serializedResult": "NWAn4ZIF",
  },
  "sampleOddRowWin": {
    "score": "Win",
    "serializedResult": "PWADOUmEANDAA",
  },
}
`;
